baseband/mark4/base.py

```diff
@@ -376,6 +376,9 @@ def read(self, count=None, fill_value=0., out=None):
             if frame_nr != self._frame_nr:
                 self._read_frame()
 
+            # Set decoded value for invalid data.
+            self._frame.invalid_data_value = fill_value
+            # Decode data into array.
             data = self._frame.data
```

Comment:
We can make the decoding somewhat less inefficient by using the fact that one can slice the frame, in which case only part will be decoded; so remove this one and rewrite as
```
nsample = ...
selection = slice(sample_offset, sample_offset + nsample)
if self.thread_ids:
    selection = (selection, self.thread_ids)
sample = self.offset - offset0
result[sample:sample + nsample] = self._frame[selection]
```

Reply:
Let's leave this for a follow-up PR (probably while looking at performance in general).